src/expr/src/scalar/like_pattern.rs

```diff
@@ -74,3 +158,774 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
+                            return false;
+                        }
+                        found = rest[..(end - 1)].rfind(&self.suffix);
+                    }
+                }
+            }
+        }
+        // No string search needed, we just use a prefix match on rest.
+        if !rest.starts_with(&self.suffix) {
+            return false;
+        }
+        rest = &rest[self.suffix.len()..];
+        match &self.next {
+            None => rest.len() == 0,
+            Some(matcher) => matcher.is_match(rest),
+        }
+    }
+}
+
+/// Breaks a LIKE pattern into a chain of sub-patterns.
+fn build_string_matcher(pattern: &str, escape: char) -> Result<StringMatcher, EvalError> {
+    let mut head = StringMatcher {
+        consume: 0,
+        many: false,
+        suffix: String::from(""),
+        next: None,
+    };
+    let mut tail = &mut head;
+    let mut in_wildcard = true;
+    let mut in_escape = false;
+
+    for c in pattern.chars() {
+        match c {
+            c if !in_escape && c == escape => {
+                in_escape = true;
+                in_wildcard = false;
+            }
+            '_' if !in_escape => {
+                if in_wildcard {
+                    tail.consume += 1;
+                } else {
+                    // start a new partial match
+                    tail.next = Some(Box::new(StringMatcher {
+                        consume: 1,
+                        many: false,
+                        suffix: String::from(""),
+                        next: None,
+                    }));
+                    tail = tail.next.as_deref_mut().unwrap();
+                    in_wildcard = true;
+                }
+            }
+            '%' if !in_escape => {
+                if in_wildcard {
+                    tail.many = true;
+                } else {
+                    // start a new partial match
+                    tail.next = Some(Box::new(StringMatcher {
+                        consume: 0,
+                        many: true,
+                        suffix: String::from(""),
+                        next: None,
+                    }));
+                    tail = tail.next.as_deref_mut().unwrap();
+                    in_wildcard = true;
+                }
+            }
+            c => {
+                tail.suffix.push(c);
+                in_escape = false;
+                in_wildcard = false;
+            }
+        }
+    }
+    if in_escape {
+        return Err(EvalError::UnterminatedLikeEscapeSequence);
+    }
+    Ok(head)
+}
+
+// Unit Tests
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use ore::str::StrExt;
+
+    #[test]
+    fn test_like() {
+        struct Input<'a> {
+            haystack: &'a str,
+            matches: bool,
+        }
+        struct Pattern<'a> {
+            needle: &'a str,
+            case_insensitive: bool,
+            escape: char,
+            inputs: Vec<Input<'a>>,
+        }
+
+        let test_cases = vec![
```

Comment:
FYI, I moved most of these into the sqllogictests in PR #10451.

Reply:
This looks great! This is the pattern that seems to work well: leave a few unit tests around to speed up the development cycle, but not so many around that it's painful to adjust them if we refactor the code under test. The SLTs have the wonderful property that they test the stable (and often ISO-standardized) interface of Materialize, so it's pretty much guaranteed that once they're written they never need to chagne.